### PR TITLE
fix large tensor softmax inplace hang

### LINKED ISSUE #45200: Softmax: Large Tensor Kernel Doesn't Support In-Place Operation
A TT_FATAL was added in https://github.com/tenstorrent/tt-metal/pull/23130/changes#diff-bf9242eb1d0a3d9b8603503f94599e212c78cb2587e8812244e3eae1631bfacfR148 that prevents in place softmax from being run with the large tensor kernels, with the error message
`Tensor is too large to run softmax inplace, please use standard softmax`

This assertion is now found here https://github.com/tenstorrent/tt-metal/blob/5e9f894fc64029da97ef0223c566f9ad91ecfac0/ttnn/cpp/ttnn/operations/normalization/softmax/device/softmax_program_factory_attention_optimized.cpp#L126

The large tensor kernel should in theory be able to function if the output memory location is the same as the input memory location. However, removing this assertion and attempting to exercise the following code results in a hang on the first iteration on an n150.

```
@pytest.mark.parametrize("target_sequence_size", [8192])
@pytest.mark.parametrize("sequence_size", [384])
@pytest.mark.parametrize("iterations", [50])
def test_transformer_attention_softmax_inplace_large_kernel_stress(
    target_sequence_size, sequence_size, iterations, device
):
    torch.manual_seed(0)

    input_shape = (1, 1, sequence_size, target_sequence_size)

    torch_input_tensor = torch_random(input_shape, -100.0, 100.0, dtype=torch.bfloat16)
    torch_attention_mask = torch_random(
        (1, 1, sequence_size, target_sequence_size), 0, 1.0, dtype=torch.bfloat16
    )

    golden_function = ttnn.get_golden_function(ttnn.transformer.attention_softmax_)
    torch_output_tensor = golden_function(torch_input_tensor, head_size=None, attention_mask=torch_attention_mask)

    # Mask is read-only across iterations, so upload once.
    attention_mask = ttnn.from_torch(torch_attention_mask, device=device, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT)

    first_output = None
    for i in range(iterations):
        print(f"[stress] iter {i}/{iterations} starting", flush=True)

        # Re-upload the input every iteration because attention_softmax_ mutates it in place.
        input_tensor = ttnn.from_torch(
            torch_input_tensor, device=device, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT
        )
        output_tensor = ttnn.transformer.attention_softmax_(
            input_tensor, head_size=None, attention_mask=attention_mask, causal_mask=True
        )
        output_tensor = ttnn.to_torch(ttnn.from_device(output_tensor))
        assert_with_pcc(torch_output_tensor, output_tensor, 0.99)

        # Check if results are ND
        if first_output is None:
            first_output = output_tensor
        else:
            assert torch.equal(first_output, output_tensor), (
                f"In-place large-kernel softmax is non-deterministic at iteration {i}: "
                f"output differs from iteration 0 (max abs delta "
                f"{(first_output - output_tensor).abs().max().item()})"
            )
```

### PR DESCRIPTION
**Pull request by ign-msati**
## PR Category

Bug fix — TTNN softmax. 
 #45200

## Summary

In-place softmax on tensors large enough to hit the attention **large-tensor kernel** was blocked by a `TT_FATAL(!attributes.inplace, ...)` assertion; removing it hung.

**Root cause** : in `softmax_large_tensor.cpp`, the fused `apply_fused_scale_mask` / `apply_fused_attn_mask` calls were passed the full capped CB length `cb_length_t` instead of the per-chunk `cur_cb_length_t` (which every other op in the loop uses). When `Wt` isn't a multiple of the capped length (e.g. `W=8192`→`Wt=256`, cap `80`, last chunk `16`), this over-consumed the input/mask CBs on the final chunk, desynced the pipeline, and deadlocked dispatch. The assertion was masking this bug, which also affected the non-in-place masked large path.

### Changes

1. `softmax_large_tensor.cpp` — pass `cur_cb_length_t` (not `cb_length_t`) to the fused scale/mask calls in all three passes (max / sum / final), matching every other op in the loop.
2. `softmax_program_factory_attention_optimized.cpp` — remove the now-unnecessary `TT_FATAL(!attributes.inplace, ...). With the kernel correct, true in-place is safe`: the compute pipeline guarantees each input tile is read before the writer overwrites it (verified on device).
3. `test_softmax.py` — add a regression test mirroring the issue's repro (in-place causal attention_softmax_, (1,1,384,8192), 50 iterations) asserting no-hang, in-place aliasing, PCC vs golden, and bit-exact determinism.

### Notes for reviewers

- Fix is the kernel one-liner (×6 sites); the assertion removal just unblocks it. No changes to output-spec/aliasing logic, so small-kernel and sharded in-place are untouched.
- In-place aliasing is safe: the reader→compute→writer CB chain enforces read-before-write per tile (verified on N150 — output aliases input, no hang, bit-identical across 50 iters).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/45200-fix-softmax-large-inplace-hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/45200-fix-softmax-large-inplace-hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ign/45200-fix-softmax-large-inplace-hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/45200-fix-softmax-large-inplace-hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/45200-fix-softmax-large-inplace-hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/45200-fix-softmax-large-inplace-hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/45200-fix-softmax-large-inplace-hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/45200-fix-softmax-large-inplace-hang)

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/fused/test_softmax.py [hunk 1]
```diff
@@ -771,3 +771,55 @@ def test_softmax_large_kernel_mask_padded(device, shape, dim):
         ulp_threshold=15,
         check_ulp=True,
     )
+
+
+@pytest.mark.parametrize("target_sequence_size", [8192])
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/fused/test_softmax.py]: Please move this test to the nightly variant of unit tests and put back the larger size that was removed in https://github.com/tenstorrent/tt-metal/pull/23130/changes/53b7c8909ce75469c785ed0fec97eef4c723d154
```
(1, 128, 128000, -1),
```
to test_large_softmax

Comment (ign-msati) [reply tests/ttnn/unit_tests/operations/fused/test_softmax.py]: done

**Review by bbradelTT — APPROVED** [review]
